### PR TITLE
feat: adds finalizers to DSC and DSCI

### PR DESCRIPTION
**Pull request by bartoszmajsak**
### **Finalizer Implementation for DSC and DSCI**
- **Based on:** PRs #301 and #605.
- Provides a default no-op `Cleanup` method for each component.
- Utilizes `controllerutil` methods for finalizer management.

### **Additional Changes:**
- Optimizes the DSCI controller: Now uses the `List` operation directly, mirroring the DSC controller's behavior and eliminating the previous `Get` then `List` approach.

## How Has This Been Tested?

In prow we trust.

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: controllers/dscinitialization/dscinitialization_controller.go [hunk 1]
```diff
@@ -63,36 +67,47 @@ type DSCInitializationReconciler struct {
...
+		if !controllerutil.ContainsFinalizer(instance, finalizerName) {
+			r.Log.Info("Adding finalizer for DSCInitialization", "name", instance.Name, "namespace", instance.Namespace, "finalizer", finalizerName)
+			controllerutil.AddFinalizer(instance, finalizerName)
+			if err := r.Update(ctx, instance); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	} else {
+		r.Log.Info("Finalization DSCInitialization start deleting instance", "name", instance.Name, "namespace", instance.Namespace, "finalizer", finalizerName)
+		// Add cleanup logic here
```
Comment (bartoszmajsak) [inline controllers/dscinitialization/dscinitialization_controller.go:99]: 🤔 I'm uncertain about the necessity of this feature as it currently appears to be a dummy implementation.

❓@VaishnaviHire @zdtsw Do you anticipate any requirement for a `Cleanup` logic specific to DSCI in upcoming iterations? If not, happy to delete this code entirely.

**Review by zdtsw — APPROVED** [review]
/lgtm

**Comment by bartoszmajsak** [comment]
I'm looking at e2e test failures

**Comment by zdtsw** [comment]
> I'm looking at e2e test failures

something is not right today for all the e2e.
i doubt it is the env, which is causing when disable dashboard it does not work.

**Review by VaishnaviHire — APPROVED** [review]
CI failures are not related to PR changes.  Tested successfully with custom image

./lgtm

**Comment by VaishnaviHire** [comment]
/approve cancel

I saw couple of errors when running e2e tests locally. Deletion test failed - 
```
    --- FAIL: TestOdhOperator/delete_components (144.80s)
        --- FAIL: TestOdhOperator/delete_components/e2e-test (140.30s)
            --- PASS: TestOdhOperator/delete_components/e2e-test/Deletion:_DataScienceCluster_instance (0.10s)
            --- FAIL: TestOdhOperator/delete_components/e2e-test/Deletion:_Application_Resource (140.20s)
FAIL
FAIL    github.com/opendatahub-io/opendatahub-operator/v2/tests/e2e     415.985s
```

with error
```
    dsc_deletion_test.go:30: 
                Error Trace:    /Users/vhire/opendatahub/opendatahub-operator/tests/e2e/dsc_deletion_test.go:30
                Error:          Received unexpected error:
                                error deleting component: kserve
                Test:           TestOdhOperator/delete_components/e2e-test/Deletion:_Application_Resource
                Messages:       Error to delete component

```

In DSC conditions, I get following error and instance in `Terminating` phase.
Screenshot 2023-10-24 at 3 00 25 PM

I had the Servicemesh operator installed, however on kserve component deletion, it deleted the openshift-operators namespace. Looking at kserve manifests

**Comment by bartoszmajsak** [comment]
> I had the Servicemesh operator installed, however on kserve component deletion, it deleted the Servicemesh subscription. Looking at kserve manifests

@VaishnaviHire that looks unrelated to me. Can you reproduce it running the latest code from `incubation` (target branch)?

**Pull request merged**